### PR TITLE
Fix DH_check() excessive time with oversized modulus

### PR DESCRIPTION
**Pull request by skmcgrail**
### Description of changes: 
Addresses CVE-2023-3446 which was low severity issue reported to OpenSSL on July 13th 2023. (See [bulletin](https://www.openssl.org/news/secadv/20230719.txt).

Excessively long DH keys can be slow to check which may lead to denial of service.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

File: crypto/fipsmodule/dh/check.c [hunk 1]
```diff
@@ -60,6 +60,7 @@
 
 #include "internal.h"
 
+#define OPENSSL_DH_CHECK_MAX_MODULUS_BITS  32768
```
Comment (justsmth) [inline crypto/fipsmodule/dh/check.c:63]: Is there a reason that `32768` was chosen?  Why not smaller? Why not larger?

The bulletin you linked only mentions 10,000:
```
Trying to use
a very large modulus is slow and OpenSSL will not normally use a modulus which
is over 10,000 bits in length.
```

Comment (skmcgrail) [reply crypto/fipsmodule/dh/check.c:63]: The 10,000 bits is the size OpenSSL limits for generation currently, but they chose to use a larger number for checking, as keys not generated by OpenSSL may have been potentially larger.

Comment (skmcgrail) [reply crypto/fipsmodule/dh/check.c:63]: >The DH_check() function checks numerous aspects of the key or parameters
that have been supplied. Some of those checks use the supplied modulus
value even if it is excessively large.
>
>There is already a maximum DH modulus size (10,000 bits) over which
OpenSSL will not generate or derive keys. DH_check() will however still
perform various tests for validity on such a large modulus. We introduce a
new maximum (32,768) over which DH_check() will just fail.
>
>An application that calls DH_check() and supplies a key or parameters
obtained from an untrusted source could be vulnerable to a Denial of
Service attack.
>
>The function DH_check() is itself called by a number of other OpenSSL
functions. An application calling any of those other functions may
similarly be affected. The other functions affected by this are
DH_check_ex() and EVP_PKEY_param_check().

The original openssl commit message details explains better.

File: crypto/fipsmodule/dh/check.c [hunk 2]
```diff
@@ -125,6 +126,13 @@ int DH_check(const DH *dh, int *out_flags) {
   BIGNUM *t1 = NULL, *t2 = NULL;
 
   *out_flags = 0;
+
+  /* Don't do any checks at all with an excessively large modulus */
```
Comment (samuel40791765) [inline crypto/fipsmodule/dh/check.c:130]: Tiny tiny nit for our coding style
```suggestion
  // Don't do any checks at all with an excessively large modulus
```

**Review by samuel40791765 — APPROVED** [review]

**Review by torben-hansen — APPROVED** [review]
Also note we use 64 rounds of MR, not 128.

**Review by justsmth — APPROVED** [review]

**Review by samuel40791765 — APPROVED** [review]

**Review by justsmth — APPROVED** [review]

**Review by samuel40791765 — APPROVED** [review]

**Pull request merged**